### PR TITLE
Refactored STP Logic

### LINKED ISSUE #1109: Clean up Play and Robot assignment code
### Description of the task

Several functions that handle tactic and play assignment in `stp.cpp` have become quite long and would likely benefit from being split up further. In particular `assignRobotsToTactics` should be broken up, perhaps to handle the goalie separately.

May be able to fix #1107 at the same time.

### Acceptance criteria

- [ ] Break up the monolithic functions in `stp.cpp` if possible.

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* Split off nongoalie tactics handling from assignRobotsToTactics
* Simplified Play::getTactics to return an empty vector if nothing to return
* Moved "static" functions to stp_impl

### Testing Done
No regressions

### Resolved Issues
resolves #1109

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/stp_impl.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * The order of the given tactics determines their priority, with the tactics as the
+ * beginning of the vector being a higher priority than those at the end. The priority
+ * determines which tactics will NOT be assigned if there are not enough robots on the
+ * field to assign them all. For example, if a Play returned 6 Tactics but there were
+ * only 4 robots on the field at the time, only the first 4 Tactics in the vector
+ * would be assigned to robots and run.
+ *
+ * @param world The state of the world for calculating robot costs
+ * @param non_goalie_robots The non goalie robots to assign to tactics
```
Comment (kelvinkoon) [inline src/software/ai/hl/stp/stp_impl.h]: Nit: Did you mean to say "The non goalie robots to assign tactics to" instead?

**Review by kelvinkoon — DISMISSED** [review]
LGTM! :+1:

File: src/software/ai/hl/stp/stp_impl.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp_impl.h]: no, we assign robots to tactics, not tactics to robots

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Overall looks good. Only real comment is why not make the `stp_impl` functions static functions of `STP`? At the very least, the `impl` suffix makes me think it's part of the [Pimpl design patterN](https://en.cppreference.com/w/cpp/language/pimpl), which it's not.

**Comment by jonathanlew** [comment]
> Overall looks good. Only real comment is why not make the `stp_impl` functions static functions of `STP`? At the very least, the `impl` suffix makes me think it's part of the [Pimpl design patterN](https://en.cppreference.com/w/cpp/language/pimpl), which it's not.

i did it along the lines of projection and projection_impl in geom, but yeah static is simpler

**Review by MathewMacDougall — APPROVED** [review]
LGTM

**Review by kelvinkoon — APPROVED** [review]

**Pull request merged**